# Data types

In Python , everything is an object and every object has a type. There are different ways to divide data types. For example You can divide them into *Scalers* which you can't sub-divide to further pieces(like a number) and non-scalers which you can sub-divide (like a list of numbers).

Another way to categorize them is based on their origin:

- Built-in Data Types
- Data types that are not built into the interpreter

Here we mainly talk about the built-in data types which can be divided into the following types:

<img src="../Images/data_types.jpeg">

Notice that this isn’t how [Python documentation](https://docs.python.org/2/library/stdtypes.html) divides data types but in my opinion the official documentation may seem a bit confusing for a reader which is facing this argument for the first time.

## Booleans

Booleans are are another core data type of Python which consist of only two values: **True** and **False**. Boolean values are mainly used when we're somehow comparing different values together. (More on comparisons later).
In Python True has a value of 1 while False's value is 0. We can check this fact by converting them to integer:

In [1]:
print(int(True))
print(int(False))

1
0


## Numeric Types


The easiest data type to work with are “Numbers” category, since all of us has a good grasp of it and already know what we should expect to see when someone talks about numbers!

Two main sub-types of numbers which we see together are:

- Integers
- Floats

(Two other numeric sub-types are long and complex which we won’t discuss here)

### Integers

Integers are numbers without decimal points:

In [2]:
1 + 1

2

In [3]:
f = 1 + 1

In [4]:
type(f)

int

In [5]:
type(1 + 1)

int

In [6]:
1256

1256

In [7]:
type(34572765 * 6779_625_645)

int

You can easily check if a number is integer or not. The only thing you need to do is use **_type( )_** built-in function (more on built-in functions later) . The result should be **_int_** as you see below:

In [8]:
type(452347)

int

In [9]:
type(1_456_099)

int

### Floats

In Python language numbers with decimal are known as Floats:

In [10]:
3.14159

3.14159

In [11]:
type(3.14159)

float

In [12]:
type('I am not a cat')

str

In [13]:
type(768.90)

float

### Integer and Float operations

Let's do some basic arithmatic and see the result data types:

In [14]:
a = 1_869 + 67.34
print(a)
print(type(a))

1936.34
<class 'float'>


In [15]:
a = 1 * 67.34
print(a)
print(type(a))

67.34
<class 'float'>


In [16]:
a = 12.5 + 10.5
print(a)
print(type(a))

23.0
<class 'float'>


In [17]:
a = 12. / 2
print(a)
print(type(a))

6.0
<class 'float'>


To summarize:
- **int** (+ or - or \*) **int** = **int**
- **float** (+ or - or \*) **float** = **float**
- **int** / **int** = **float**
- **int** / **float** = **float**
- **float** / **float** = **float**

### Integer and Float conversion (Cast)

Using **_float( )_** and **_int( )_** we can convert integers to floats and vice versa:

In [18]:
val = 34.67
type(val)
int(val)
print(val)

34.67


In [19]:
print(int(45.89)) # convert float to int
print(int('347')) # convert string to int

45
347


In [20]:
"""if you want to convert a string to int, the string MUST have
the form of an integer"""
print(int('56.76')) # this one doesn't work!

ValueError: invalid literal for int() with base 10: '56.76'

In [ ]:
print(int(float('56.76'))) # this one does, why?

56


In [ ]:
# The same thing is not true for converting string to float.
# it means the conversion works even if the string doesn't have the form of a 
# float
float('4')

4.0

<img src="../Images/student.svg"   width="30" align="left">               

**YOUR TURN :** Considering **a = 3.14**, which code results in 3 ?

- A : floor(a)
- B: float(int(a))
- C : int(10 // a)
- D : None of above